src/include/daos_pipeline.h

```diff
@@ -0,0 +1,261 @@
+#define DAOS_FILTER_AKEY		201
+#define DAOS_FILTER_CONST		202
+
+/** types used in a filter object */
+#define DAOS_FILTER_TYPE_BINARY		0
+#define DAOS_FILTER_TYPE_STRING		1
+#define DAOS_FILTER_TYPE_INTEGER	2
+#define DAOS_FILTER_TYPE_REAL		3
+
+/** types of pipeline nodes */
+#define DAOS_PIPELINE_CONDITION		0
+#define DAOS_PIPELINE_AGGREGATION	1
+
+
+/**
+ * A filter object, used to build operations for a pipeline node.
+ *
+ */
+typedef struct {
+	/**
+	 * filter type can be any of the following:
+	 *   -- functions:
+	 *      - logical functions:
+	 *          DAOS_FILTER_FUNC_EQ:		==
+	 *          DAOS_FILTER_FUNC_NE:		!=
+	 *          DAOS_FILTER_FUNC_LT:		<
+	 *          DAOS_FILTER_FUNC_LE:		<=
+	 *          DAOS_FILTER_FUNC_GE:		>=
+	 *          DAOS_FILTER_FUNC_GT:		>
+	 *          DAOS_FILTER_FUNC_LIKE:		== (reg exp.)
+	 *          DAOS_FILTER_FUNC_ISNULL:		==NULL
+	 *          DAOS_FILTER_FUNC_ISNOTNULL:		!=NULL
+	 *          DAOS_FILTER_FUNC_AND:		&&
+	 *          DAOS_FILTER_FUNC_OR:		||
+	 *      - aggeration functions:
+	 *          DAOS_FILTER_FUNC_SUM:		SUM()
+	 *          DAOS_FILTER_FUNC_MIN:		MIN()
+	 *          DAOS_FILTER_FUNC_MAX:		MAX()
+	 *          DAOS_FILTER_FUNC_AVG:		AVG()
+	 *   -- key:
+	 *          DAOS_FILTER_DKEY:	Filter object represents dkey
+	 *          DAOS_FILTER_AKEY	Filter object represents akey
+	 *   -- constant:
+	 *          DAOS_FILTER_CONST:	Filter object is a constant
+	 */
+	int		filter_type;
+	/**
+	 * Type of data. Only relevant for keys and constant filter type objects:
+	 *          DAOS_FILTER_TYPE_BINARY
+	 *          DAOS_FILTER_TYPE_STRING
+	 *          DAOS_FILTER_TYPE_INTEGER
+	 *          DAOS_FILTER_TYPE_REAL
+	 */
+	int		data_type;
+	/**
+	 * Number of parameters for this filter object. For example, for '=='
+	 * we have 2 parameters.
+	 */
+	uint32_t	num_params;
+	/**
+	 * If filtering by akey, this tells us which one.
+	 */
+	d_iov_t		akey;
+	/**
+	 * How many constants we have in \a constant
+	 */
+	size_t		num_constants;
+	/**
+	 * This object holds the value of the constants
+	 */
+	d_iov_t		*constant;
+	/**
+	 * If filter should only be applied starting at an offset of the data.
+	 */
+	size_t		data_offset;
+	/**
+	 * Size of the data to be filtered.
+	 */
+	size_t		data_len;
+} daos_pipeline_filter_t;
+
+/**
+ * A pipeline node, used to build a pipeline.
+ */
+typedef struct {
+	/**
+	 * Node type can be any of the following:
+	 *   -- DAOS_PIPELINE_CONDITION:
+	 *          Records in, and records (meeting condition) out
+	 *   -- DAOS_PIPELINE_AGGREGATION:
+	 *          Records in, a single value out
+	 *
+	 * NOTE: Pipeline nodes can only be chained the following way:
+	 *             (condition) --> (condition)
+	 *             (condition) --> (aggregation)
+	 */
+	int			node_type;
+	/**
+	 * Number of filters inside this pipeline node
+	 */
+	size_t			num_filters;
+	/**
+	 * Array of filters for this node
+	 */
+	daos_pipeline_filter_t	*filters;
+} daos_pipeline_node_t;
```

Comment:
My original idea what to call this a pipeline and the next level above this a dataflow. Or maybe call this a pipe .... That being said, I am ok to use your semantic if there is consensus about it.

Reply:
Pipeline requires a set of steps, and each step is a node. A node has filters. In reality, a node is just one filter, but this filter is composed of filter objects. If this is confusing, we can change the names of the types.

To understand this, let say that you have a pipeline that does filtering first, and aggregation second. In this case you have a pipeline with 2 nodes. The first node is a filtering node, and the second is the aggregation node. Now, the filtering node filters by, let say, "(name LIKE 'A%') AND (age<40)", and the aggregation does SUM(age). In the first node, you have a logical expression composed of 7 filter objects. In the second, you have an aggregation expression composed of 2 filter objects:

**node1:**  | AND | LIKE | akey=name | const='A%' | < | akey=age | const=40 |
**node2:**  | SUM | akey=age |

Does this make sense to you? A node does a single filtering, although the filtering may be complex and composed of multiple "filter" objects.

Reply:
This will be renamed as daos_filter_t